### PR TITLE
Oauth Provider tests

### LINKED ISSUE #2040: Authorization: argument of type 'NoneType' is not iterable
Since the addition of https://github.com/MIT-LCP/physionet-build/pull/2003, we have been regularly seeing an `argument of type 'NoneType' is not iterable` error:

```
Internal Server Error: /content/autonomic-aging-cardiovascular/

TypeError at /content/autonomic-aging-cardiovascular/
argument of type 'NoneType' is not iterable

Traceback:

File "/physionet/python-env/physionet/lib/python3.9/site-packages/django/core/handlers/exception.py" in inner
 56.                 response = get_response(request)

File "/physionet/python-env/physionet/lib/python3.9/site-packages/oauth2_provider/middleware.py" in __call__
 32.                 user = authenticate(request=request)

File "/physionet/python-env/physionet/lib/python3.9/site-packages/django/views/decorators/debug.py" in sensitive_variables_wrapper
 42.             return func(*func_args, **func_kwargs)

File "/physionet/python-env/physionet/lib/python3.9/site-packages/django/contrib/auth/__init__.py" in authenticate
 77.             user = backend.authenticate(request, **credentials)

File "./user/backends.py" in authenticate
 15.         if '@' in username:

Exception Type: TypeError at /content/autonomic-aging-cardiovascular/
Exception Value: argument of type 'NoneType' is not iterable
Request information:
USER: AnonymousUser

GET: No GET data
POST: No POST data
FILES: No FILES data
COOKIES: No cookie data
```

The error can be reproduced with:

```
curl -H user-agent:'Microsoft Office Word 2014' -H accept-auth:'badger,Wlid1.1,Bearer,Basic,NTLM,Digest,Kerberos,Negotiate,Nego2' -H authorization:'Bearer' localhost:8000
```

See also: https://github.com/MIT-LCP/physionet-build/pull/2037.  For now a temporary fix is: https://github.com/MIT-LCP/physionet-build/pull/2039

@bemoody's suggestions/notes on this issue:

1. Our custom authorization (`DualAuthModelBackend`) does not seem to be behaving in the way that Django expects it to behave. We should look at the current authorization module and compare to our implementation.
2. It is possible that the fix in https://github.com/MIT-LCP/physionet-build/pull/2039 will break OAuth.
3. We need to better understand the OAuth module and work out what it is doing. We could consider temporarily removing the OAuth middleware in the meantime.

### PR DESCRIPTION
**Pull request by Rutvikrj26**
As Described by @bemoody in #2040, I've added the Test Case for OAuth. This test case allows access to a protected resource endpoint on the server. 

The Test first sets up an OAuth application and generates a token. This is the BaseTest, which is the parent and gets inherited for further tests. There are two testcases we want to set up - as mentioned in #2040. The first testcase checks the ability to access the protected resource once a user has a token. The second testcase focuses on authentication and achieving the token that can allow access. 

Testcase 1 : Once the person has a token, he can get authorized to access Hello - an endpoint emulating a protected resource. There are two tests added - One demonstrating a denial for unauthenticated access. The second one has an auth header with a bearer token and is allowed to access the protected resource. 

Testcase 2 : This test demonstrated the authorization code flow for two different kind of applications supported by OAuth Workflow. 

1. Confidential Application: OAuth2 client credentials is a grant flow that allows a web service or an app (confidential client) to use its own credentials, instead of impersonating a user, to authenticate when calling another web service or a web resource, such as a REST API. This is commonly used for server-to-server interactions that run in the background without immediate user interaction. The permissions are granted directly to the application itself by an administrator.

2. Public Applciation: A Public Application can not securely store client secrets, like a mobile application. These are untrusted and hence follow a [Proof Key for Code Exchange (PKCE) workflow](https://auth0.com/docs/get-started/authentication-and-authorization-flow/authorization-code-flow-with-proof-key-for-code-exchange-pkce).

There are two tests, each resembling one application type and the respective authentication flow to be used for implementing the authentication.

As can be seen, we can implement a direct channel between Physionet and Health Data Nexus, which is not exposed to the regular public. An authenticated user on Health Data Nexus can easily access a protected resource on Physionet(for Eg: Training), since the access will be granted to the HDN platform.

### CODE REVIEW

File: physionet-django/oauth/views.py [hunk 1]
```diff
@@ -2,6 +2,6 @@
 from oauth2_provider.views.generic import ProtectedResourceView
 
 
-class ApiEndpoint(ProtectedResourceView):
+class hello(ProtectedResourceView):
```
Comment (tompollard) [inline physionet-django/oauth/views.py:5]: Did this happen intentionally?

Comment (Rutvikrj26) [reply physionet-django/oauth/views.py:5]: Yup - I changed it to Hello to keep the view name and the URL endpoint have the same name.

**Comment by bemoody** [comment]
Sounds good - thanks!

It looks like your test cases aren't currently using the 'applications/register/' view, so is there a reason that you moved it?

(I don't know if we wanted to limit the use of OAuth, at least initially, by making those views only accessible in debug mode.  I'm not opposed in principle to opening the system up to anyone who wants to use it, but we'd need to understand the implications.)

**Comment by Rutvikrj26** [comment]
Thanks for catching the Register Endpoint!
That one is supposed to be inside the debug and only be available via the Django Superuser Panel in production. Once the team agrees on the implementation, I can add these endpoints to the Admin panel, but until we start using Oauth, it is better not to have them in production.

**Comment by bemoody** [comment]
This looks good, thanks!

One small nit - do you need to define `get_random_string()` here?  Couldn't you use `from django.utils.crypto import get_random_string` instead?

I think we also want to test the token returned in `test_basic_auth` and `test_secure_auth_pkce`, by doing something like:
```
        ...
        response = self.client.post(
            reverse("oauth2_provider:token"), data=token_request_data, **auth_headers
        )
        self.assertEqual(response.status_code, 200)

        self.client.logout()

        token = response.json()['access_token']
        auth = self._create_authorization_header(token)
        response = self.client.get("/oauth/hello", HTTP_AUTHORIZATION=auth)
        self.assertEqual(response.status_code, 200)
```
(or perhaps we don't need to call `client.login` to begin with, since we're using basic auth?)

**Comment by bemoody** [comment]
Thanks!

**Pull request merged**